server/api.go

```diff
@@ -495,6 +505,116 @@ func (p *Plugin) getYourPrs(w http.ResponseWriter, r *http.Request) {
 	w.Write(resp)
 }
 
+func (p *Plugin) getPrsDetails(w http.ResponseWriter, r *http.Request) {
+	userID := r.Header.Get("Mattermost-User-ID")
+	if userID == "" {
+		http.Error(w, "Not authorized", http.StatusUnauthorized)
+		return
+	}
+
+	ctx := context.Background()
+
+	var githubClient *github.Client
+
+	if info, err := p.getGitHubUserInfo(userID); err != nil {
+		writeAPIError(w, err)
+		return
+	} else {
+		githubClient = p.githubConnect(*info.Token)
+	}
+
+	var prList []*PRDetails
+	json.NewDecoder(r.Body).Decode(&prList)
+
+	var prDetails []*PRDetails
+
+	ch := make(chan *PRDetails)
+	for _, pr := range prList {
+		go fetchPRDetails(ctx, githubClient, pr.URL, pr.Number, ch)
```

Comment:
To clarify the flow of data here, I propose that instead of passing a channel, just have `fetchPRDetails` return `(*PRDetails, error)`, and then coordinate here instead. This decouples `fetchPRDetails` from how a caller wants to use it,  and simplifies the return conditions of that code. 

On this side, the code then might look something like:

```go
prDetails = make([]*PRDetails, len(prList))
var wg sync.WaitGroup
for i, pr := range prList {
    i := i
    pr := pr
    wg.Add(1)
    go func() {
        defer wg.Done()
        prDetail, err := fetchPRDetails(ctx, githubClient, pr.URL, pr.Number)
        if err != nil {
            // log error ...
            prDetails[i] = ... // default value?
            return
        }
        
        prDetails[i] = prDetail
    }()
}
wg.Wait()

resp, _ := json.Marshal(prDetails)
w.Write(resp)
```

Reply:
Thanks @lieut-data! I can see the improvement of doing things your way. I had understood that "the way to do things" in go was by using channels ("do not communicate by sharing memory, share memory by communication") but this simplify some things. Nevertheless, I do not see so much improvement on the fetchPRDetails conditions, so I am not sure if I am missing something. Nevertheless, I just pushed my changes for when you have time to take a look at them.

Reply:
That maxim is indeed idiomatic, but its thrust is to avoid writing to shared memory from multiple processes which can be fraught with race conditions. The specific advice that applies here comes from https://github.com/golang/go/wiki/CodeReviewComments#synchronous-functions:
> Prefer synchronous functions - functions which return their results directly or finish any callbacks or channel ops before returning - over asynchronous ones.

> Synchronous functions keep goroutines localized within a call, making it easier to reason about their lifetimes and avoid leaks and data races. They're also easier to test: the caller can pass an input and check the output without the need for polling or synchronization.

> If callers need more concurrency, they can add it easily by calling the function from a separate goroutine. But it is quite difficult - sometimes impossible - to remove unnecessary concurrency at the caller side.

Thanks for iterating on this -- I'll take a look at your changes soon :)